### ISSUE TITLE
load stupidest possible circuits as TD6 files to see what the game will/won't load

### ISSUE THREAD

**Issue by kevinburke** [kevinburke/rct#3]
experiments:

- short complete track
- incomplete track
- complete track with collision
- complete track with invalid piece

**Comment by kevinburke** [comment]
I'm fairly certain it won't load a track with collisions. Not sure about the other cases.

**Comment by crhallberg** [comment]
I generated a simple track by preventing the track from changing height (blacklisted all tracks with `_TO_` in the name). I got a "Cannot construct ride unless closed" message. Is there a ride open/closed flag in the TD6 file?

**Comment by kevinburke** [comment]
Interesting. It's possible? I have been able to load tracks in the game with the following code (this is from the server that's since been abandoned)

```
	return func(w http.ResponseWriter, r *http.Request) {
		iterPath := strings.Replace(r.URL.Path, "/td6", "", 1)
		m, err := loadMember(directory, iterPath)
		if err != nil {
			w.WriteHeader(http.StatusBadRequest)
			w.Write([]byte(err.Error()))
			return
		}
		query := r.URL.Query()
		completeParam := query.Get("complete")
		complete := completeParam == "true"
		// For whatever reason RCT2 needs the extension to be in uppercase
		shortName := fmt.Sprintf("%s.TD6", truncate(17, m.Id))
		w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", shortName))
		w.Header().Set("Content-Type", "text/td6")
		ride := td6.CreateMineTrainRide(m.Track, complete)
		for i := 0; i < len(ride.TrackData.Elements); i++ {
			fmt.Println(tracks.ElementNames[ride.TrackData.Elements[i].Segment.Type])
		}
		bits, err := td6.Marshal(ride)
		paddedBits := td6.Pad(bits)
		rleWriter := rle.NewWriter(w)

		if err != nil {
			w.WriteHeader(http.StatusBadRequest)
			w.Write([]byte(err.Error()))
			return
		}
		w.WriteHeader(http.StatusOK)
		rleWriter.Write(paddedBits)
	}
```

**Comment by kevinburke** [comment]
The important part of that I think is td6.CreateMineTrainRide

**Comment by kevinburke** [comment]
This is the openrct2 definition of a ride: https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/ride/ride.h#L132-L332

**Comment by crhallberg** [comment]
Yes, I used the server to get my test file after altering the experiment. Having slept on it, the track was not a loop, so I wonder if "closed" didn't mean "switched off" like when you're playing, but a complete loop. I'm going to try to download a completed track and try again.

**Comment by crhallberg** [comment]
After a lot of tinkering, I have not only have a 3D preview working in the game, I remembered how to take screenshots! However, I'm still getting the error (second screenshot). I still think it's a flag issue of some kind (looking at the definition now). I have a suspicion that it may have to do with the entrances and exits. Onward!

![scr1](https://cloud.githubusercontent.com/assets/451107/20240223/0c7c64fa-a8e1-11e6-8332-f8633ed56721.jpg)

And that error I mentioned before.
![scr2](https://cloud.githubusercontent.com/assets/451107/20240227/22ae3816-a8e1-11e6-8a8d-1c20415c4583.jpg)

**Comment by crhallberg** [comment]
SO! I noticed the turn collision only as I posted the image here. I *JUST* tweaked my track completion code (I rewrote yours since it wasn't completing tracks correctly when I tried) and now...

Previews in track selection:
![scr3](https://cloud.githubusercontent.com/assets/451107/20240282/8867afba-a8e2-11e6-822b-0ceeeb8c7a2a.jpg)

AND THIS:
![scr4](https://cloud.githubusercontent.com/assets/451107/20240285/9181ee62-a8e2-11e6-9c2a-f7160d20559b.jpg)

So I spoke too soon in my last update. Sorry. And yay!

**Comment by kevinburke** [comment]
Great! Yeah I've found if I can't get something running the easiest way to
figure out why is to build each track piece individually in the game and
find the one that won't build.
